src/plugins/tonegen/ausrv.c

```diff
@@ -0,0 +1,332 @@
+
+You should have received a copy of the GNU Lesser General Public
+License along with this library; if not, write to the Free Software
+Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301
+USA.
+*************************************************************************/
+
+#define _GNU_SOURCE
+
+#include <stdlib.h>
+#include <string.h>
+#include <errno.h>
+
+#include <ngf/log.h>
+#include <trace/trace.h>
+
+#include "stream.h"
+#include "ausrv.h"
+
+#if PA_API_VERSION < 9
+#error Invalid PulseAudio API version
+#endif
+
+#define DISCONNECTED    0
+#define CONNECTED       1
+
+#define DEFAULT_SERVER  "default Pulse Audio"
+#define CONNECT_DELAY   10                              /* in seconds */
+
+#define TRACE(f, args...) trace_write(trctx, trflags, trkeys, f, ##args)
+#define LOG_CAT "tonegen-ausrv: "
+
+static void set_connection_status(struct ausrv *, int);
+static void context_callback(pa_context *, void *);
+static void event_callback(pa_context *, pa_subscription_event_type_t,
+                           uint32_t, void *);
+static void connect_server(struct ausrv *);
+static void restart_timer(struct ausrv *, int);
+static void cancel_timer(struct ausrv *);
+
+static const char *pa_client_name = "ngf-tonegen-plugin";
+
+
+int ausrv_init(void)
+{
+    return 0;
+}
+
+void ausrv_exit(void)
+{
+}
+
+struct ausrv *ausrv_create(struct tonegend *tonegend, char *server)
+{
+    pa_glib_mainloop   *mainloop = NULL;
+    struct ausrv       *ausrv;
+    pa_mainloop_api    *mainloop_api;
+
+    if ((ausrv = malloc(sizeof(*ausrv))) == NULL) {
+        N_ERROR(LOG_CAT "%s(): Can't allocate memory", __FUNCTION__);
+        goto failed;
+    }
+
+    if ((mainloop = pa_glib_mainloop_new(NULL)) == NULL) {
+        N_ERROR(LOG_CAT "%s(): pa_glib_mainloop_new() failed", __FUNCTION__);
+        goto failed;
+    }
+
+    mainloop_api = pa_glib_mainloop_get_api(mainloop);
+
+    if (pa_signal_init(mainloop_api) < 0) {
+        N_ERROR(LOG_CAT "%s(): pa_signal_init() failed", __FUNCTION__);
+        goto failed;
+    }
+    
+    memset(ausrv, 0, sizeof(*ausrv));
+    ausrv->tonegend = tonegend;
+    ausrv->server   = strdup(server ? server : DEFAULT_SERVER);
+    ausrv->mainloop = mainloop;
+
+    connect_server(ausrv);
+
+    return ausrv;
+
+ failed:
+    if (mainloop != NULL)
+        pa_glib_mainloop_free(mainloop);
+
+    if (ausrv != NULL)
+        free(ausrv);
+
+    return NULL;
+
+}
+
+void ausrv_destroy(struct ausrv *ausrv)
+{
+    if (ausrv != NULL) {
+        stream_kill_all(ausrv);
+
+        if (ausrv->context != NULL)
+            pa_context_unref(ausrv->context);
+        
+        if (ausrv->mainloop != NULL)
+            pa_glib_mainloop_free(ausrv->mainloop);
+        
+        free(ausrv->server);
+        free(ausrv);
+    }
+}
+
+
+static void set_connection_status(struct ausrv *ausrv, int sts)
+{
+    int connected = sts ? CONNECTED : DISCONNECTED;
+
+    if ((ausrv->connected ^ connected)) {
```

Comment:
Again, I'd use `bool` for the status and then just use `true` and `false` and `!=` instead of xor here.